dev_setup.sh

```diff
@@ -0,0 +1,113 @@
+#!/usr/bin/env bash
+
+# cryptol setup script
+#
+# Supported distribution(s):
+#  * macOS 14 (AArch64)
+#
+# This script installs everything needed to get a working development
+# environment for cryptol. Any new environment requirements should be
+# added to this script.
+#
+
+set -e
+
+HERE=$(cd `dirname $0`; pwd)
+LOG=$HERE/dev_setup.log
+
+function notice {
+    echo "[NOTICE] $*"
+}
+
+# Requires: LOG set to log file path.
+function logged {
+    if ! [ -z "$LOG" ]
+    then
+        mkdir -p `dirname $LOG`
+        echo $* >>$LOG
+        if ! $* >>$LOG 2>&1
+        then
+            echo
+            echo "An error occurred; please see $LOG"
+            echo "Here are the last 50 lines:"
+            tail -n 50 $LOG
+            exit 1
+        fi
+    fi
+}
+
+function update_submodules {
+    cd $HERE
+    notice "Updating submodules"
+    git submodule update --init
+}
+
+function install_ghcup {
+    if ! ghcup --version &> /dev/null
+    then
+        notice "Installing ghcup, GHC, and cabal"
+        # technically the installation only requires cabal, but it's 
+        # recommended to get the whole GCH shebang in one package
+        curl --proto '=https' --tlsv1.2 -sSf https://get-ghcup.haskell.org | sh
+    else
+        notice "Using existing ghcup installation"
+    fi
+
+}
+
+# Indicate whether this is running macOS on the Apple M series hardware.
+function is_macos_aarch {
+    # Is it running macOS?
+    if [[ "$OSTYPE" != "darwin"* ]]; then
+        notice "dev setup does not currently support your OS / hardware"
+        exit 1 
+    fi
+
+    # Does it use an M-series chip?
+    chip=$(system_profiler SPHardwareDataType | grep Chip)
+    [[ $chip == *M1* || $chip == *M2* || $chip == *M3* ]]
+}
+
+function install_gmp {
+    if is_macos_aarch; then
+        notice "Installing gmp via Homebrew, if it's not already installed"
+        logged brew list gmp || brew install gmp
+
+        # on macOS 12 (x86_64), I think homebrew uses different locations
+        notice "You may need to add the following environment variables to your '.profile': \n"\
```

Comment:
This is a slightly bigger-picture comment, but I would recommend putting these environment variables in, say, `$repo_root/env.sh` and then having those sourced as a matter of course when working with the project. My thinking behind this suggestion is to centralize all environment requirements within the physical confines of the repo rather than asking the user to update (say) `~/.profile`, which can be easy to forget as an integral part of getting a working system set up.

Reply:
I suggest this partly to encourage standardization of essential settings (and people could always have local files to customize those). Mainly, to avoid, hat in hand, having to tell the user to set variables just to get basic project tasks done. But also, I'm assuming that such an `env.sh` can be written to capture all the assumptions about repo structure, which is then hopefully stable across all development environments and can be maintained by the whole team. If you don't have that, then my `.profile` can become stale because of a change someone makes to a script in the repo.

Reply:
(To summarize it a little more concisely, I like to find ways to make sure that the repo owns the definition of what a complete setup looks like and how it works as one source of truth, since spreading out that truth over both the repo and local files can result in builds that should work but don't, or do but shouldn't.)

Reply:
I can't say I disagree really; it's just that unless you only support exactly identically cloned environments, there's some amount of tree config that is going to be (in general) different for everyone. I guess the actual issue is that it shouldn't be done with environment variables.

In my world, there's a place in the tree that is not managed by source control where you put your local tree config, and it's structured so that if you don't have anything there you get reasonable defaults, and otherwise there's a reference for what you can set and what the pertinent considerations are. Basically this evolved from the premise that anything should build with ```./configure && make``` with the observation that well-written (and thus non-autoconf) configure scripts produce an output file that a clueful user can equally well manage by hand. But it relies on having build systems that are adequately expressive to capture this kind of stuff in a non-hacky way. E.g. for this issue in one of my own projects the config would probably be "put ```GMP_PREFIX=/opt/homebrew``` in src/config.mk". I don't think there's a good way to get cabal to do this kind of thing (but I'd be happy to wrong about that!) so it becomes a tradeoff between what's nice and what's reasonably expedient.

Reply:
@sauclovian-g I'm not really sure what to say to that in response; while I don't want to question the merits of your approach and past experiences, I would still like my original request here to stand on the grounds that the environment variables being mentioned are, it would seem, a requirement for working with the project and so they strike me as the sort of thing you would put with the project. I'm also hoping we can get to some resolution on this to avoid confusion for @marsella if there's conflicting messages in this discussion.

Reply:
I added this but it's a little messy -- repeated runs of the script will re-append the same two lines repeatedly. Would it be reasonable to delete & re-create that file each time we run the setup script or do we want to account for the possibility that someone will separately put their own environment vars in there?

Reply:
@marsella I'm not sure what you're referring to (I don't see any new commits). Can you say a bit more about what you mean by "re-append the same two lines" and "re-create that file"?

Reply:
Looking at the latest changes, I see that `env.sh` is being *generated* by `dev_setup.sh`, so it sounds like above you're saying that is what's messy (appending to the generated file). I had assumed that `env.sh` would be static and put under version control, not generated.

Reply:
@jtdaugherty we aren't in my world, was partly wandering off into the weeds and partly trying to sort out what we do need. It seems that the build system we have needs environment variables set, and changing that (if it's possible at all) is well beyond the scope of this ticket. But the general points about tree config remain.

ISTM that there are basically two choices:
1. check in a separate env.sh that probes the system on the fly and sets the environment variables (with two subcases, one where you run it with ```. ./env.sh``` and one where you run it with ```eval $(./env.sh)```)
2. have the setup script generate env.sh based on what it finds, akin to a configure script

of which (1) is possibly more robust but (2) is simpler and closer to what's already here. So in the interests of not dragging this out or making extra work, I think I'd suggest (2). I think the difference from what we already have is basically just writing out a new env.sh each time instead of appending...

Reply:
however if @marsella is up for writing (1) I have no objection to that either...

I guess a third option might be to check in a separate env.sh for each environment but that has diminishing returns...

Reply:
@sauclovian-g either option works for me and I think addresses the concern that I was trying to express, namely, just not to ask the user to do manually what could otherwise be automated to increase consistency.

Reply:
Right

Reply:
Thank you for all this discussion! I took option 2.

> [...] the environment variables being mentioned are, it would seem, a requirement for working with the project

This particular instance is more about setting up Homebrew correctly, which is independent of this particular project and which others may have already done correctly or differently on their own machines. E.g. you can change the default prefix that brew uses, so I would not want to hard-code the environment variables in a file because it could bork someone's system that doesn't use the default Homebrew options, which aren't necessary to run this project. I think that's what @sauclovian-g's option 1 is getting at.